### PR TITLE
add term aggregation

### PR DESCRIPTION
**Pull request by PSeitz**

### CODE REVIEW

File: src/aggregation/agg_req_with_accessor.rs [hunk 1]
```diff
@@ -27,11 +31,32 @@ impl AggregationsWithAccessor {
     }
 }
 
+#[derive(Clone)]
+pub(crate) enum FastFieldAccessor {
+    Multi(MultiValuedFastFieldReader<u64>),
+    Single(DynamicFastFieldReader<u64>),
+}
+impl FastFieldAccessor {
+    pub fn as_single(&self) -> &DynamicFastFieldReader<u64> {
```
Comment (fulmicoton) [inline src/aggregation/agg_req_with_accessor.rs]: can we return an Option and panic on the call site?

File: src/aggregation/agg_req.rs [hunk 2]
```diff
@@ -102,10 +102,25 @@ pub(crate) struct BucketAggregationInternal {
 impl BucketAggregationInternal {
     pub(crate) fn as_histogram(&self) -> &HistogramAggregation {
```
Comment (fulmicoton) [inline src/aggregation/agg_req.rs]: can we return an Option and panic on the call site?

File: src/aggregation/agg_req.rs [hunk 3]
```diff
@@ -102,10 +102,25 @@ pub(crate) struct BucketAggregationInternal {
...
         match &self.bucket_agg {
-            BucketAggregationType::Range(_) => panic!("unexpected aggregation"),
             BucketAggregationType::Histogram(histogram) => histogram,
+            _ => panic!("unexpected aggregation"),
         }
     }
+    pub(crate) fn as_term(&self) -> &TermsAggregation {
```
Comment (fulmicoton) [inline src/aggregation/agg_req.rs]: can we return an Option and panic on the call site?

File: src/aggregation/agg_req_with_accessor.rs [hunk 4]
```diff
@@ -27,11 +31,32 @@ impl AggregationsWithAccessor {
...
+        match self {
+            FastFieldAccessor::Multi(_) => panic!("unexpected ff cardinality"),
+            FastFieldAccessor::Single(reader) => reader,
+        }
+    }
+    pub fn as_multi(&self) -> &MultiValuedFastFieldReader<u64> {
```
Comment (fulmicoton) [inline src/aggregation/agg_req_with_accessor.rs]: can we return an Option and panic on the call site?

File: src/aggregation/agg_req_with_accessor.rs [hunk 5]
```diff
@@ -121,29 +158,36 @@ pub(crate) fn get_aggs_with_accessor_and_validate(
 fn get_ff_reader_and_validate(
     reader: &SegmentReader,
     field_name: &str,
-) -> crate::Result<(DynamicFastFieldReader<u64>, Type)> {
+    multi: bool,
```
Comment (fulmicoton) [inline src/aggregation/agg_req_with_accessor.rs]: I don't understand the multi argument here. 

What happens if multi == true and Cardinality is single valued?
Is this ok?

Can we comment on the function?
Shouldl the bool be of the type Cardinality instead? and call it cardinality too?

**Review by fulmicoton — CHANGES_REQUESTED** [review]
See minor commetns inline

File: src/aggregation/agg_req.rs (hunk 3 above)
Comment (PSeitz) [reply src/aggregation/agg_req.rs]: yep

File: src/aggregation/agg_req_with_accessor.rs (hunk 5 above)
Comment (PSeitz) [reply src/aggregation/agg_req_with_accessor.rs]: I refactored the function and it accepts cardinality as parameter, I think it should be clear now

File: src/aggregation/mod.rs [hunk 6]
```diff
@@ -337,17 +338,79 @@ mod tests {
...
+                Term::from_field_text(text_field, term),
+                IndexRecordOption::Basic,
+            );
+
+            searcher.search(&term_query, &collector)?
+        } else {
+            searcher.search(&AllQuery, &collector)?
+        };
+
+        let res: Value = serde_json::from_str(&serde_json::to_string(&agg_res)?)?;
```
Comment (fulmicoton) [inline src/aggregation/mod.rs:367]: Can you add a commment explaining why we serialize and deserialize?
Wouldn't serde_json::to_value do the job ?

File: src/aggregation/mod.rs [hunk 7]
```diff
@@ -337,17 +338,79 @@ mod tests {
...
+        agg_req: Aggregations,
+        index: &Index,
+        query: Option<(&str, &str)>,
+    ) -> crate::Result<Value> {
+        let collector = AggregationCollector::from_aggs(agg_req);
+
+        let reader = index.reader()?;
+        let searcher = reader.searcher();
+        let agg_res = if let Some((field, term)) = query {
+            let text_field = reader.searcher().schema().get_field(field).unwrap();
```
Comment (fulmicoton) [inline src/aggregation/mod.rs:354]: can we remove this unwrap?

I don't understand why it is justified. Assuming it is justified, that's a hidden contract.

Comment (fulmicoton) [reply src/aggregation/mod.rs:354]: ah ok this is just for tests! don't worry then!

File: src/aggregation/bucket/range.rs [hunk 8]
```diff
@@ -102,22 +103,69 @@ pub struct SegmentRangeCollector {
     field_type: Type,
 }
 
+#[derive(Clone, PartialEq)]
+pub(crate) struct SegmentRangeBucketEntry {
+    pub key: Key,
+    pub doc_count: u64,
+    pub sub_aggregation: Option<SegmentAggregationResultsCollector>,
+    /// The from range of the bucket. Equals f64::MIN when None.
+    pub from: Option<f64>,
```
Comment (fulmicoton) [inline src/aggregation/bucket/range.rs:112]: what kind of interval is this? Semi-open?

File: src/aggregation/agg_req_with_accessor.rs (hunk 5 above)
Comment (fulmicoton) [reply src/aggregation/agg_req_with_accessor.rs]: Is this pushed?

Comment (PSeitz) [reply src/aggregation/agg_req_with_accessor.rs]: It should be now

File: src/aggregation/mod.rs [hunk 9]
```diff
@@ -454,17 +533,18 @@ mod tests {
...
         } else {
             let collector = AggregationCollector::from_aggs(agg_req);
 
             let searcher = reader.searcher();
-            searcher.search(&term_query, &collector).unwrap()
+            searcher.search(&AllQuery, &collector).unwrap()
         };
 
         let res: Value = serde_json::from_str(&serde_json::to_string(&agg_res)?)?;
+        // println!("{}", serde_json::to_string_pretty(&res).unwrap());
```
Comment (fulmicoton) [inline src/aggregation/mod.rs]:
```suggestion
```

File: src/aggregation/agg_req_with_accessor.rs [hunk 10]
```diff
@@ -78,10 +115,14 @@ impl MetricAggregationWithAccessor {
...
             | MetricAggregation::Stats(StatsAggregation { field: field_name }) => {
-                let (accessor, field_type) = get_ff_reader_and_validate(reader, field_name)?;
+                let (accessor, field_type) =
+                    get_ff_reader_and_validate(reader, field_name, Cardinality::SingleValue)?;
 
                 Ok(MetricAggregationWithAccessor {
-                    accessor,
+                    accessor: accessor
+                        .as_single()
+                        .expect("unexpected fast field cardinatility")
```
Comment (fulmicoton) [inline src/aggregation/agg_req_with_accessor.rs]:
```suggestion
                        .expect("unexpected fast field cardinality")
```

File: src/aggregation/mod.rs (hunk 6 above)
Comment (PSeitz) [reply src/aggregation/mod.rs:367]: yes, serde_json::to_value is better, I changed it

Comment (PSeitz) [reply src/aggregation/mod.rs:367]: ah no, I used that here to also test de/serialization roundtrip

**Review by fulmicoton — APPROVED** [review]

**Pull request merged**